pywbem_mock/_baserepository.py

```diff
@@ -0,0 +1,353 @@
+# version 2.1 of the License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful, but
+# WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+# Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public
+# License along with this program; if not, write to the Free Software
+# Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
+#
+# Author: Karl  Schopmeyer <inovadevelopment.com>
+#
+
+
+"""
+Base classes for an object store for CIM classes, CIM instances, and CIM
+qualifier declarations and the generic API for a repository of these object
+types organized by repository.
+"""
+from abc import ABCMeta, abstractmethod, abstractproperty
+
+import six
+
+
+def compatibleabstractproperty(func):
+    """
+    Python 2 and python 3 differ in decorator for abstract property.
+    in python 3 (gt 3.3) it is:
+        @property
+        @abstractproperty
+    in python 2
+        @abstractproperty
+    """
+
+    if six.PY3:
+        return property(abstractmethod(func))
+    else:
+        return abstractproperty(func)
+
+
+@six.add_metaclass(ABCMeta)
+class ObjectStoreAPI(object):
+    """
+    This abstract class defines the APIs for the methods of an object store
+    for CIM objects including CIMClass, CIMInstance, and CIMQualifierDeclaration
+    objectsthat constitute a WBEM server repository.  This
+    class provides the abstract methods for creating, accessing, and deleting,
+    CIM objects of a single CIM object type in the repository.
+
+    These APIs allows creating, updating, deleting, and retrieving these
+    objects.
+    """
+
+    def __init__(self, case_insensitive_names, cim_object_type):
+        """
+        Initialize the object store.
+
+        Parameters:
+
+            case_insensitive-names(:class:`py:bool`):
+              If True, the names uniquely identify objects in the object store
+              case insensitively. The names must be strings.
+              If False, names identify objects. The names may be strings
+              or other objects that can be used to uniquely identify
+              objects in the object store (ex. CIMInstanceName).
+
+            cim_object_type(:term:`string`):
+              The Pywbem cim object type as defined in cim_types.py for the
+              objects in the data store. Used to verify values on create.
+        """
+        self._case_insensitive_names = case_insensitive_names
+        self._cim_object_type = cim_object_type
+
+    @abstractmethod
+    def exists(self, name):
+        """
+        Test if cim_object defined by name exists in the
+        object store
+
+        Paramsters
+          name(:term:`string` or :class:`~pywbem.CIMInstanceName`):
+            Name by which the object is identified in the object store
```

Comment:
Here we are very specific in what the type can be besides a string. In the description of the `case_insensitive_names` init parameter, the name could be any object. Decide for one approach and use that consistently.

Reply:
CLOSED - Comment has become moot, because the parameter has been removed.